xmipptomo/protocols/protocol_coords_roi.py

```diff
@@ -26,68 +26,112 @@
-                if abs(coorm.getX() - coorc.getX()) <= distance and abs(coorm.getY() - coorc.getY()) <= distance \
-                        and abs(coorm.getZ() - coorc.getZ()) <= distance:
-                    outputSet.append(coorc)
-        # selection options??
-        self._defineOutputs(outputCoordinates=outputSet)
-        self._defineSourceRelation(inputCoor, outputSet)
+        sel = self.selection.get()
+        for ix, inputSetCoor in enumerate(self.inputCoordinates):
+            perc = self._percentage(inputSetCoor)
+            for mesh in self.inputMeshes.get().iterItems():
+                if inputSetCoor.get().getPrecedents().getFirstItem().getFileName() == mesh.getVolume().getFileName():
+                    if sel == 0:
+                        i = 0
+                    else:
+                        outputSetList = []
+                    for coorcc in inputSetCoor.get():
+                        for coormesh in mesh.getMesh():
+                            if self._euclideanDistance(coorcc, coormesh) <= self.distance.get():
+                                if sel == 0:
+                                    i += 1
+                                else:
+                                    outputSetList.append(coorcc.getObjId())
+                                break
+                    if sel == 0:
+                        if i >= perc:
+                            outputSet = self._createSetOfCoordinates3D(inputSetCoor.get().getPrecedents(), ix + 1)
+                            outputSet.copyInfo(inputSetCoor)
+                            outputSet.copyItems(inputSetCoor.get())
+                            outputSet.setBoxSize(inputSetCoor.get().getBoxSize())
+                            outputSet.setSamplingRate(inputSetCoor.get().getSamplingRate())
+                            name = 'output3DCoordinates%s' % str(ix+1)
+                            args = {}
+                            args[name] = outputSet
+                            self._defineOutputs(**args)
+                            self._defineSourceRelation(inputSetCoor, outputSet)
+
+                    else:
+                        if len(outputSetList) != 0:
+                            outputSet = self._createSetOfCoordinates3D(inputSetCoor.get().getPrecedents(), ix + 1)
+                            outputSet.copyInfo(inputSetCoor)
+                            outputSet.setBoxSize(inputSetCoor.get().getBoxSize())
+                            outputSet.setSamplingRate(inputSetCoor.get().getSamplingRate())
+                            for coor3D in inputSetCoor.get().iterItems():
+                                if coor3D.getObjId() in outputSetList:
+                                    outputSet.append(coor3D)
+                            name = 'output3DCoordinates%s' % str(ix + 1)
+                            args = {}
+                            args[name] = outputSet
+                            self._defineOutputs(**args)
```

Comment:
Defining output on the fly...is working almost ok. We need to be careful when running things on parallel and see how this behaves. 

What we need to do, at least here is to flag the sets created as "open" and at the end, (in a final step....close the sets.

This is something scipion might take care in the future but for now you need to do it yourself.

Reply:
@fede-pe is well aware of what I'm saying. ;-)

Reply:
Every protocol from scipion-em-imod works this way. I would suggest to take a look to newstack protocol (method getOutputInterpolatedSetOfTiltSeries). It is the simplest one and it will illustrate what @pconesa is saying. Also, to properly close the set, there is an example of how to do it in the fiducialModel protocol in the createOuputStep method.

Reply:
thanks @fede-pe